Week1/index.js

```diff
@@ -0,0 +1,173 @@
+    }
+    console.log(`Records added successfully into ${table} table.`);
+  } catch (err) {
+    if (err) {
+      showErrorAndExit(`Query Error: ${err.message} (${err.code})`, connection);
+    }
+  }
+}
+
+// Populate tables with related data
+async function readAndInsertDataIntoTables() {
+  try {
+    // Read & Parse Data
+    const { cities, countries } = JSON.parse(await readFile('./data.json', 'utf-8'));
+
+    // Insert each city record to the city table
+    await insertRecord(cities, 'city');
+
+    // Insert each country record to the country table
+    await insertRecord(countries, 'country');
+
+    // Before exit, close the connection
+    connection.end();
+  } catch (error) {
+    showErrorAndExit(`File Read Error: ${error.message}`, connection);
+  }
+}
+
+async function queryAndShowResultOf(filter, table, fields) {
+  const query = `select ${
+    fields.length > 1 ? fields.join(',') : fields[0]
+  } from ${table} ${filter}`;
+  try {
+    const result = await asyncQuery(query);
+    console.log('\n\nResult of ', filter, '\n');
+    console.table(result);
+  } catch (err) {
+    showErrorAndExit(`Filtering Error on (${filter}): ${err.message}`, connection);
+  }
+}
+
+/**
+ * Just to prevent duplicates
+ * At first run it should be 'false'
+ * Then it must be 'true' for the purpose
+ */
+const isTablePopulatedBefore = false;
+
+async function main() {
+  // Questions 1 - 3
+  await createDBAndTables();
+  if (!isTablePopulatedBefore) await readAndInsertDataIntoTables();
+
+  // Question 4
+  // 1 - names of the countries with population greater than 8 million
+  await queryAndShowResultOf('where population > 8000000', 'country', ['name']);
+  // 2 - names of the countries that have “land” in their names
+  await queryAndShowResultOf(`where name like '%land%'`, 'country', ['name']);
+  // 3 - names of the cities with population in between 500,000 and 1 million
+  await queryAndShowResultOf('where population between 500000 and 1000000', 'city', ['name']);
+  // 4 - names of all the countries on the continent ‘Europe’
+  await queryAndShowResultOf(`where continent = 'Europe'`, 'country', ['name']);
+  // 5 - all the countries in the descending order based on their surface areas
+  await queryAndShowResultOf(`order by surface_area desc`, 'country', ['*']);
+
+  // Question 5 => Optionals
+  // 1- names of all the cities in the Netherlands?
+  await queryAndShowResultOf(`where country_code = 'NLD'`, 'city', ['*']);
+  // 2 - population of Rotterdam
+  await queryAndShowResultOf(`where name = 'Rotterdam'`, 'city', ['population']);
+  // 3 - top 10 countries based on surface area
+  await queryAndShowResultOf(`order by surface_area desc limit 10`, 'country', ['*']);
+  // 4 - top 10 cities with the highest population
+  await queryAndShowResultOf(`order by population desc limit 10`, 'city', ['*']);
+  // 5 - population of the world
+  await queryAndShowResultOf('', 'country', ['sum(population) as world_population']);
```

Comment:
Your queries are all good!

Reply:
Thank you very much for your kind explanations and review :)
I tried to avoid repetition for the sake of homework but I agree that sometimes avoiding repetitions makes code less readable :(